source/mesh_deformation/interface.cc

```diff
@@ -0,0 +1,639 @@
+    {
+      // Interpolate the mesh vertex velocity onto the Stokes velocity system for use in ALE corrections
+      LinearAlgebra::BlockVector distributed_mesh_velocity;
+      distributed_mesh_velocity.reinit(sim.introspection.index_sets.system_partitioning, sim.mpi_communicator);
+
+      const std::vector<Point<dim> > support_points
+        = sim.finite_element.base_element(sim.introspection.component_indices.velocities[0]).get_unit_support_points();
+
+      Quadrature<dim> quad(support_points);
+      UpdateFlags update_flags = UpdateFlags(update_values | update_JxW_values);
+      FEValues<dim> fs_fe_values (*sim.mapping, free_surface_fe, quad, update_flags);
+      FEValues<dim> fe_values (*sim.mapping, sim.finite_element, quad, update_flags);
+      const unsigned int n_q_points = fe_values.n_quadrature_points,
+                         dofs_per_cell = fe_values.dofs_per_cell;
+
+      std::vector<types::global_dof_index> cell_dof_indices (dofs_per_cell);
+      FEValuesExtractors::Vector extract_vel(0);
+      std::vector<Tensor<1,dim> > velocity_values(n_q_points);
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = sim.dof_handler.begin_active(), endc= sim.dof_handler.end();
+      typename DoFHandler<dim>::active_cell_iterator
+      fscell = free_surface_dof_handler.begin_active();
+
+      for (; cell!=endc; ++cell, ++fscell)
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (cell_dof_indices);
+
+            fe_values.reinit (cell);
+            fs_fe_values.reinit (fscell);
+            fs_fe_values[extract_vel].get_function_values(fs_mesh_velocity, velocity_values);
+            for (unsigned int j=0; j<n_q_points; ++j)
+              for (unsigned int dir=0; dir<dim; ++dir)
+                {
+                  unsigned int support_point_index
+                    = sim.finite_element.component_to_system_index(/*velocity component=*/ sim.introspection.component_indices.velocities[dir],
+                                                                                           /*dof index within component=*/ j);
+                  distributed_mesh_velocity[cell_dof_indices[support_point_index]] = velocity_values[j][dir];
+                }
+          }
+
+      distributed_mesh_velocity.compress(VectorOperation::insert);
+      mesh_velocity = distributed_mesh_velocity;
+    }
+
+
+    template <int dim>
+    void MeshDeformationHandler<dim>::setup_dofs()
+    {
+      if (!sim.parameters.free_surface_enabled)
+        return;
+
+      // these live in the same FE as the velocity variable:
+      mesh_velocity.reinit(sim.introspection.index_sets.system_partitioning,
+                           sim.introspection.index_sets.system_relevant_partitioning,
+                           sim.mpi_communicator);
+
+
+      free_surface_dof_handler.distribute_dofs(free_surface_fe);
+
+      sim.pcout << "Number of free surface degrees of freedom: "
```

Comment:
get this through the simulator access base class

Reply:
So it's safer/better to use the getters when available, even though we have direct access to the simulator?

Reply:
I think so. I don't particularly like the fact that we're having `friend` classes and would want to eventually get rid of that. Not directly accessing variables is going to make it easier later on to achieve that goal.